src/Shared/UnitTests/ObjectModelHelpers.cs

```diff
@@ -1817,7 +1817,12 @@ private static string[] SplitPathIntoFragments(string path)
         /// </summary>
         internal static void DeleteFiles(params string[] paths)
         {
-            foreach (string path in paths)
+            // When we delete the file directory which has the sub folder/file firstly, it will not be delected since not empty.
+            // So sort paths descendingly by depth, it will delete sub folder/file at first.
+            var pathsSortedByDepth = paths.OrderByDescending(x =>
+                                       x.NormalizeForPathComparison().Split(Path.DirectorySeparatorChar).Length);
```

Comment:
nitpick: the `NormalizeForPathComparison` is probably needed only for trimming the trailing slashes (as it's unlikely there will be mismatched slashes here - as rest of the function would not work either), and triming of trailing slash would not be needed if we'd ignore empty results

```suggestion
                                       x.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries).Length);
```

Alternatively we can just sort by the path length - though the depth of nessting might vary here, it will still ensure we are deleting content of subdir before moving to parent dir (as the subdir substring will always be of a same length).

Reply:
On this codepath it doesn't matter, since it's test cleanup, but `string.Split` is very allocatey. Here, to count slashes in the string, the code has to allocate (for each path):

1. A string for every path element.
2. An array to hold pointers to those strings

And all that becomes garbage at the end of the line, since it's no longer reachable after `Length` returns.

Again, this would be important in a hot codepath and isn't very important here: none of the allocated things will be on the large-object heap, so they'll get collected in the next gen0 collection, which is fast. But our codebase does have many areas where this kind of thing matters so it's good to think about.

Reply:
> Alternatively we can just sort by the path length - though the depth of nessting might vary here, it will still ensure we are deleting content of subdir before moving to parent dir (as the subdir substring will always be of a same length).

 Sorting by the path length. There is one condition that parent file name is much logger than the subfolder and name.

> nitpick: the `NormalizeForPathComparison` is probably needed only for trimming the trailing slashes (as it's unlikely there will be mismatched slashes here - as rest of the function would not work either), and triming of trailing slash would not be needed if we'd ignore empty results

I use this function NormalizeForPathComparison since the ToPlatformSlash() in it.  It will normalize the slashes.

Reply:
> There is one condition that parent file name is much longer than the subfolder and name together.

Right - but this can only happen for _unrelated_ paths, correct?
E.g. `C:\loooooooongpath`, `C:\a\b\c.txt`, `C:\a\b\d.txt` - here there is no harm to process the `C:\loooooooongpath` last. We only need to make sure to process all subpaths for particular path before the path itself (to make sure it's empty) - and that will allways happen, as related paths have common prefix and hence subpaths must be longer then prents (unless we are using soft/hard links - but then all other path comparing approachs are doomed).
Or is there specific scenario where we'll fail to delete something when comparing just based on lengths?

Reply:
Yeah. We can sort by file directory length.  I'm thinging the file path including file name.  Have updated.

Reply:
I was actually thinking about the full path, including the filename, as well

```
var pathsSortedByDepth = paths.OrderByDescending(x => x.Length);
```

Is there a scenario where content of the directory would not be deleted before moving to parent drectory (unless relative and absolute paths are mixed together)

Reply:
> I was actually thinking about the full path, including the filename, as well

If including filename， parent filename length are greater than subfolder name+ filename length.